core/src/main/java/com/datastax/oss/driver/internal/core/ssl/ReloadingKeyManagerFactory.java

```diff
@@ -0,0 +1,257 @@
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.X509ExtendedKeyManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ReloadingKeyManagerFactory extends KeyManagerFactory implements AutoCloseable {
+  private static final Logger logger = LoggerFactory.getLogger(ReloadingKeyManagerFactory.class);
+  private static final String KEYSTORE_TYPE = "JKS";
+  private Path keystorePath;
+  private String keystorePassword;
+  private ScheduledExecutorService executor;
+  private final Spi spi;
+
+  // We're using a single thread executor so this shouldn't need to be volatile, since all updates
+  // to lastDigest should come from the same thread
+  private volatile byte[] lastDigest;
+
+  /**
+   * Create a new {@link ReloadingKeyManagerFactory} with the given keystore file and password,
+   * reloading from the file's content at the given interval. This function will do an initial
+   * reload before returning, to confirm that the file exists and is readable.
+   *
+   * @param keystorePath the keystore file to reload
+   * @param keystorePassword the keystore password
+   * @param reloadInterval the duration between reload attempts. Set to {@link
+   *     java.time.Duration#ZERO} to disable scheduled reloading.
+   * @return
+   */
+  public static ReloadingKeyManagerFactory create(
+      Path keystorePath, String keystorePassword, Duration reloadInterval) {
+    KeyManagerFactory kmf;
+    try {
+      kmf = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+    } catch (NoSuchAlgorithmException e) {
+      throw new RuntimeException(e);
+    }
+
+    KeyStore ks;
+    try (InputStream ksf = Files.newInputStream(keystorePath)) {
+      ks = KeyStore.getInstance(KEYSTORE_TYPE);
+      ks.load(ksf, keystorePassword.toCharArray());
+    } catch (IOException | CertificateException | KeyStoreException | NoSuchAlgorithmException e) {
+      throw new RuntimeException(e);
+    }
+    try {
+      kmf.init(ks, keystorePassword.toCharArray());
+    } catch (KeyStoreException | NoSuchAlgorithmException | UnrecoverableKeyException e) {
+      throw new RuntimeException(e);
+    }
+
+    ReloadingKeyManagerFactory reloadingKeyManagerFactory = new ReloadingKeyManagerFactory(kmf);
+    reloadingKeyManagerFactory.start(keystorePath, keystorePassword, reloadInterval);
+    return reloadingKeyManagerFactory;
+  }
+
+  @VisibleForTesting
+  protected ReloadingKeyManagerFactory(KeyManagerFactory initial) {
+    this(
+        new Spi((X509ExtendedKeyManager) initial.getKeyManagers()[0]),
+        initial.getProvider(),
+        initial.getAlgorithm());
+  }
+
+  private ReloadingKeyManagerFactory(Spi spi, Provider provider, String algorithm) {
+    super(spi, provider, algorithm);
+    this.spi = spi;
+  }
+
+  private void start(Path keystorePath, String keystorePassword, Duration reloadInterval) {
+    this.keystorePath = keystorePath;
+    this.keystorePassword = keystorePassword;
+    this.executor =
```

Comment:
just occurred to be me that in theory we could avoid creating the executor (and thus a thread) if the reload interval is zero right?

Reply:
from some local testing at least from what I can tell this never creates a thread if nothing is ever scheduled, so maybe this isn't needed (fine with it either way)

Reply:
I'll avoid constructing the executor if possible, worth doing